src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,493 @@
+     * {@code ExtractThenExpand} objects.
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder ofExtract() {
+        return new Builder().createBuilder();
+    }
+
+    /**
+     * Defines the input parameters of an {@code Expand} object
+     *
+     * @param prk
+     *     the pseudorandom key; must not be {@code null}
+     * @param info
+     *     the optional context and application specific information (may be
+     *     {@code null}); the byte[] is copied to prevent subsequent
+     *     modification
+     * @param length
+     *     the length of the output key material (must be &gt; 0 and &lt; 255 *
+     *     HMAC length)
+     *
+     * @return a new {@code Expand} object
+     *
+     * @throws NullPointerException
+     *     if {@code prk} is {@code null}
+     * @throws IllegalArgumentException
+     *     if {@code length} is not > 0
+     */
+    static Expand expandOnly(SecretKey prk, byte[] info, int length) {
+        if (prk == null) {
+            throw new NullPointerException("prk must not be null");
+        }
+        return new Expand(prk, info, length);
+    }
+
+    /**
+     * Defines the input parameters of an Extract operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Extract implements HKDFParameterSpec {
+
+        // HKDF-Extract(salt, IKM) -> PRK
+        private final List<SecretKey> ikms;
+        private final List<SecretKey> salts;
+
+        private Extract(List<SecretKey> ikms, List<SecretKey> salts) {
+            this.ikms = List.copyOf(ikms);
+            this.salts = List.copyOf(salts);
+        }
+
+        /**
+         * Returns an unmodifiable {@code List} of input key material values in
+         * the order they were added.
+         *
+         * @return the unmodifiable {@code List} of input key material values
+         */
+        public List<SecretKey> ikms() {
+            return ikms;
+        }
+
+        /**
+         * Returns an unmodifiable {@code List} of salt values in the order they
+         * were added.
+         *
+         * @return the unmodifiable {@code List} of salt values
+         */
+        public List<SecretKey> salts() {
+            return salts;
+        }
+
+    }
+
+    /**
+     * Defines the input parameters of an Expand operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Expand implements HKDFParameterSpec {
+
+        // HKDF-Expand(PRK, info, L) -> OKM
+        private final SecretKey prk;
+        private final byte[] info;
+        private final int length;
+
+        /**
+         * Constructor that may be used to initialize an {@code Expand} object
+         *
+         * @param prk
+         *     the pseudorandom key; may be {@code null}
```

Comment:
Instead of stating "prk may be {@null}", how about narrow it down/clarify to "in the case of ExtractThenExpand, prk may be {@null} since the output of extract phase is used"

Reply:
Yes, resolved.